build.gradle

```diff
@@ -10,10 +10,24 @@ buildscript {
 
 apply plugin: 'com.android.library'
 
+def eclipseProperties=new Properties()
+file('project.properties').withInputStream {
+    eclipseProperties.load(it)
+}
+
 android {
     compileSdkVersion 21
     buildToolsVersion "21.1.2"
 
+    defaultConfig {
+        targetSdkVersion eclipseProperties.getProperty('target')-'android-'
+
+        // note that this code assumes it's building the latest tag, and that all tags are present in the repository;
+        // older tags can be built by cloning only a subsection of the history
+        versionCode 'git tag'.execute(null, projectDir) .text.readLines().size()
```

Comment:
Does `versionCode` even need to be kept around? I originally had it in the manifest because it was part of the boilerplate when I generated the project, but I wonder if it's only needed for applications and not libraries.

Reply:
Strictly speaking no, it doesn't. Currently the manifests of library projects are completely ignored when building an application.

However, the [consensus best practice](https://stackoverflow.com/questions/7377025/is-versioncode-versionname-required-in-library-manifest) seems to be to include the version information anyway. These attributes might not be used at the moment, but that doesn't mean they never will be, and putting them in future-proofs the library. Besides, if nothing else, having a `versionName` derived from `git describe` provides an easy way to determine the exact commit from which a binary was built.

Reply:
Sounds good. For `versionCode`, please change it to the number of commits rather than the number of tags. I think `git rev-list --count HEAD` will do the trick.

Reply:
You sure about that?  I'll make the changes if that's really what you want, but the problem with using a commit count for `versionCode` is that it isn't always guaranteed to go up.  If there's ever a release from a branch that isn't merged back into the trunk, its `versionCode` may not properly reflect the chronological release order, as it is supposed to.

Reply:
I don't anticipate ever tagging releases from a branch. And even if I did (e.g. a bug fix release for an older major/minor), I don't think they should be considered updates to what is on master. The best way to do a version code from a semantic version would probably be a formula like `versionCode = (major * 1000000) + (minor * 1000) + patch`. That's probably overkill for what is needed here though.

Reply:
Fair enough.  See the added commit for relevant updates.